### ISSUE TITLE
How do I debug an error

### ISSUE THREAD

**Issue by IngoVals** [timber/timber#1727]
So I'm trying to add Timber to a theme, without using the given startertheme. However just to get it start working and seeing some results and I'm right away having some problems.

It seems that the Render method returns nothing.

I tried this  

```
$compiled = Timber::compile( 'index.twig', $context );
var_dump($compiled);
```

And what is dumped is just `bool(false)`

I'm assuming that means that something is wrong with my setup. I'm guessing it is something very basic, but I'm not sure how I would get the output to tell me what that is.

All documentation seems to point towards only using the starter theme, and no explanation on doing your own so I guess I'm missing something that is required but never mentioned.

I assume that TimberSite class is a required one, but not sure. I tried this in `functions.php`

```
Timber::$dirname = 'views';

class RutSite extends TimberSite {

}
new RutSite();
```

And here is my index.php

```
<?php
/** 
 * 
 * @package rut
 */

if (!class_exists('Timber')){
    echo 'Timber not activated. Make sure you activate the plugin in <a href="/wp-admin/plugins.php#timber">/wp-admin/plugins.php</a>';
    return;
}

$context = Timber::get_context();
$post = new TimberPost();
$context['post'] = $post;

$compiled = Timber::compile( 'index.twig', $context );
var_dump($compiled);

//Timber::render( 'index.twig', $context );
```

Is there no way to get a stacktrace or something that could at least give me a basic clue what is going on.

**Comment by gchtr** [comment]
Hey @IngoVals 

Thanks for pointing out what you miss in the documentation or what you – as a Timber beginner – don’t understand (yet). We’re always trying to improve the documentation and try to make it work for everyone.

> All documentation seems to point towards only using the starter theme, and no explanation on doing your own so I guess I'm missing something that is required but never mentioned.

I’m afraid you’re right. A lot of documentation points to the starter theme, but it is actually not required to start out with Timber at all.

> I assume that TimberSite class is a required one, but not sure. I tried this in functions.php

You don’t need to instantiate a `Timber\Site` object. The only thing you need to do is to instantiate Timber (like in the [Setup Guide](https://timber.github.io/docs/getting-started/setup/)):

**functions.php**

```php
new Timber\Timber();
```

Your **index.php** file looks fine from what I can see. Just to make sure, when you call `$compiled = Timber::compile( 'index.twig', $context );`, Timber will look for an **index.twig** file in the **views** folder of your theme. The starter theme works different here, because it places Twig files in the **templates** folder. That means that you don’t need the statement `Timber::$dirname = 'views';` either. So my two question to boil this down are:

- Where did you put your **index.twig** file?
- What does your **index.twig** file look like?
- Are you sure your **index.php** file is called for the page you look at? Sometimes WordPress chooses a different template file than we first think, based on [what it can find](http://wphierarchy.com/). It might help if you work with the [Debug Bar Plugin](https://wordpress.org/plugins/debug-bar/) and the [Timber Debug Bar](https://wordpress.org/plugins/debug-bar-timber/) to check what WordPress and Timber are doing.

> Is there no way to get a stacktrace or something that could at least give me a basic clue what is going on.

If you don’t use an [IDE like PhpStorm for debugging](https://deliciousbrains.com/xdebug-advanced-php-debugging/) your code, I can only point you to the Debug Bar plugins mentioned above and the [Debugging Guide](https://timber.github.io/docs/guides/debugging/).

**Comment by gchtr** [comment]
When `Timber::compile()` returns `false`, it usually means that it couldn’t find the Twig file. But I currently don’t see why it wouldn’t work, based on your explanation. More questions:

- What do you see when you do a `var_dump( Timber::$locations );` just before you run `Timber::compile()`? Are there any special chars or spaces in the absolute path to the Twig file on your machine?
- Did you set `define( 'WP_DEBUG', true );` in your **wp-config.php**?
- What version of WordPress, PHP and Timber are you using?
- How did you install Timber? As a plugin, or through Composer?

**Comment by IngoVals** [comment]
Before answering your questions, I do believe I've found the cause, but not the reason.

You see I'm running WP as Docker container, I tried to move the theme result over to a local XAMPP stack and there it works.

So it's likely something to do with Docker. The routing to my php files work fine, but I guess to the twig files not so much.

So I did this `Timber::$locations = get_template_directory() .'/views';`

And now it works. So there is something about Volumes in Docker Containers that messes with locating the twig files.

Here is my docker-compose if you intend to investigate any further, but at least this works currently for me.

```
version: '3'

services:
  db:
    image: mysql:5.7
    volumes:
      - ./.data/db:/var/lib/mysql
    restart: always
    environment:
      MYSQL_ROOT_PASSWORD: somewordpress
      MYSQL_DATABASE: wordpress
      MYSQL_USER: wordpress
      MYSQL_PASSWORD: wordpress

  wordpress:
    depends_on:
      - db
    image: wordpress:latest
    ports:
      - "8000:80"
    volumes:
      - ./dist:/var/www/html/wp-content/themes/rut
      - ./wp-content/plugins:/var/www/html/wp-content/plugins
    restart: always
    environment:
      WORDPRESS_DB_HOST: db:3306
      WORDPRESS_DB_USER: wordpress
      WORDPRESS_DB_PASSWORD: wordpress
```

Notice the theme folder in the container points towards the dist folder in my project setup. I use webpack to compile the things I need and drop theme there.

I'll monitor this thread If anybody has questions for me or want me to test something out for them. I'll leave it up to you to decide if you want this closed ( Bug is not fixed, but easily circumvented by using the tools of Timber ).

Thank you for your help.

PS. FYI the `var_dump( Timber::$locations );` returned NULL before I explicitly set it. If it should be evaluated to views folder then obviously there is where something breaks.

**Comment by gchtr** [comment]
Okay, in that case I think we found where the problem lies. I guess you won’t have to answer the other questions.

I myself don’t work with Docker, so I can’t help you here. I just saw that `Timber::$locations` actually only contains values if you set them yourself. And it’s only used by the Loader which figures out the template locations if it’s not empty. I’m not too familiar with Timber’s Loader that handles all the locations, so I’m afraid I can’t hunt this down currently. Maybe @jarednova can? But I think using `Timber::$locations = get_template_directory() .'/views';` is a good workaround in this case.

Let’s leave this open for now.
